### PR TITLE
ReaderAnnotation: skip old highlights with invalid xpointers

### PR DESCRIPTION
**Pull request by hius07**
Supercedes https://github.com/koreader/koreader/pull/12493.
Closes https://github.com/koreader/koreader/issues/12492. Closes https://github.com/koreader/koreader/issues/12443.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12494)

### CODE REVIEW

**Review by Frenzie — APPROVED** [review]

**Comment by Frenzie** [comment]
@Hzj-jie Works for you?

**Comment by Hzj-jie** [comment]
I cannot verify it again since the broken xpointers got cleaned up after a successful open, though I think it should work.
But a very noticeable issue is that this change silently drops user data because of the cre change which is also something I highly discourage.

**Comment by hius07** [comment]
1) There is a warning in the log.
2) This occurs when old highlights/bookmarks are converted to annotations, the new format. Old hl/bm are not deleted, they remain in the sdr file.
3) Do you think we should show an info message?

**Comment by Hzj-jie** [comment]
I don't think regular users would read logs at all.
But are you suggesting that the changes of xpointer would only occur when converting old highlights/bookmarks to annotations?

**Comment by hius07** [comment]
The crash in the sorting func occurs when coverting old highlights only.

**Comment by Hzj-jie** [comment]
lgtm

**Comment by Hzj-jie** [comment]
Remaining in the sdr file still seems useless for regular users. But anyway, I closed the original PR.

**Comment by hius07** [comment]
> Remaining in the sdr file still seems useless for regular users.

Agree. It was done on purpose within the great refactoring of bookmarks/highlights to annotations, to avoid data losses.
It works rather good now, so we may consider removing old bm/hl after conversion, to save disk space.

File: frontend/apps/reader/modules/readerannotation.lua [hunk 1]
```diff
@@ -11,6 +11,12 @@ local ReaderAnnotation = WidgetContainer:extend{
 
 function ReaderAnnotation:buildAnnotation(bm, highlights, init)
     -- bm: associated single bookmark ; highlights: tables with all highlights
+    if self.ui.rolling and
+            (not self.document:isXPointerInDocument(bm.page) or
+            (bm.pos1 and not self.document:isXPointerInDocument(bm.pos1))) then
+        logger.warn("Invalid xpointer in old highlight, skipping")
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerannotation.lua:17]: Might be worth logging the xpointer too, so users can find it out and see what text it is about, and possibly fix them manually (or rehighlight them)?

**Review by poire-z — APPROVED** [review]

**Pull request merged**